ncm-opennebula/src/main/perl/OpenNebula/Network.pm

```diff
@@ -0,0 +1,207 @@
+                    $self->remove_duplicate_ars($one, $vnet, $vnar, $arinfo);
+                    $self->verbose("vnet: $vnet already contains AR id: ", $arinfo->{AR_ID}->[0]);
+                } else {
+                    $self->create_vn_ars($one, $vnet, $ardata, $vnar);
+                };
+            }
+       }
+    }
+}
+
+=item detect_duplicate_ars
+
+Detects duplicate C<VirtualNetwork> C<ARs> with
+same IPs or MACs.
+Removes duplicated C<ARs> (if C<QUATTOR> flag is set to true).
+
+=cut
+
+sub remove_duplicate_ars
+{
+    my ($self, $one, $vnet, $vnar, $arinfo) = @_;
+
+    # Try to find different AR ids with the same configuration
+    if (defined($arinfo)) {
+        my %arpool = $vnar->get_ar_pool();
+        my $mac = $arinfo->{MAC}->[0];
+        my $ip = $arinfo->{IP}->[0];
+        my $arid = $arinfo->{AR_ID}->[0];
+        foreach my $id (sort keys %arpool) {
+            if (($arpool{$id}->{MAC}->[0] eq $mac) && ($id != $arid)) {
+                $self->warn("Current AR $arid and $vnet AR $id have the same MAC $mac");
+                $self->remove_vn_ars($one, $arpool{$id}, $vnet, $arinfo, $vnar);
+            } elsif (($arpool{$id}->{IP}->[0] eq $ip) && ($id != $arid)) {
+                $self->warn("Current AR $arid and $vnet AR $id have the same IP $ip");
+                $self->remove_vn_ars($one, $arpool{$id}, $vnet, $arinfo, $vnar);
+            }
+        }
+    }
+}
+
+=item create_vn_ars
+
+Creates C<VirtualNetwork> C<AR> leases.
+
+=cut
+
+sub create_vn_ars
+{
+    my ($self, $one, $vnet, $ardata, $ar) = @_;
+
+    # Create a new network AR
+    $self->debug(1, "New AR template in $vnet: ", $ardata->{ar});
+    my $arid = $ar->addar($ardata->{ar});
+    if (defined($arid)) {
+        $self->info("Created new $vnet AR id: ", $arid);
+    } else {
+        $self->error("Unable to create new AR within vnet: $vnet");
+    }
+}
+
+=item remove_vn_ars
+
+Removes <VirtualNetwork> C<AR> leases.
+
+=cut
+
+sub remove_vn_ars
+{
+    my ($self, $one, $arinfo, $vnet, $ardata, $ar) = @_;
+    my $arid;
+    # Detect Quattor and id first
+    $arid = $self->detect_vn_ar_quattor($arinfo) if $arinfo;
+    if (defined($arid)) {
+        $self->debug(1, "AR template to remove from $vnet: ", $ardata->{ar});
+        my $rmid = $ar->rmar($arid);
+        if (defined($rmid)) {
+            $self->info("Removed from vnet: $vnet AR id: $arid");
+        } else {
+             $self->error("Unable to remove AR id: $arid from vnet: $vnet");
+        }
+    } elsif ($arinfo) {
+        $self->error("Unable to remove AR. Quattor flag not found within AR: $arinfo");
+    } else {
+        $self->warn(1, "Unable to remove AR. ",
+            "AR template is not available from vnet: $vnet: ", $ardata->{ar});
+    }
+}
+
+=item remove_vn_ars
+
+Detects C<Quattor> flag within C<AR> template.
+
+=cut
+
+sub detect_vn_ar_quattor
+{
+    my ($self, $ar)  =@_;
+    my $arid = $ar->{AR_ID}->[0];
+
+    if ($ar->{QUATTOR}->[0]) {
```

Comment:
if `QUATTOR` is not set, this will give a failure or warning; should be `$ar->{QUATTOR} && $ar->{QUATTOR}->[0]`